presto-main/src/main/java/com/facebook/presto/operator/ExchangeClientManager.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.memory.context.LocalMemoryContext;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public class ExchangeClientManager
+{
+    private final ExchangeClientSupplier supplier;
+    private final List<ExchangeClient> exchangeClients;
+
+    @VisibleForTesting
+    public ExchangeClientManager(ExchangeClientSupplier supplier)
+    {
+        this.supplier = requireNonNull(supplier, "supplier is null");
+        this.exchangeClients = new ArrayList<>();
+    }
+
+    public ExchangeClient createExchangeClient(LocalMemoryContext systemMemoryContext)
```

Comment:
I'm afraid it has to be `synchronized`. This method is called from the `OperatorFactory#createOperator`, and that method is not guarranteed to be called from a single thread in a sequential manner.